### ISSUE TITLE
Confusing and incomplete documentation/validator behavior about banned CSS classes and tag names

### ISSUE THREAD

**Issue by zhangsu** [ampproject/amphtml#26292]
**Please only file reports about bugs in AMP here.**

## What's the issue?

Briefly describe the bug.

> Class names, in author stylesheets, may not start with the string `-amp-`. These are reserved for internal use by the AMP runtime. It follows, that the user's stylesheet may not reference CSS selectors for `-amp-` classes and `i-amp` tags."

 https://amp.dev/documentation/guides-and-tutorials/develop/style_and_layout/style_pages/#disallowed-styles

There are multiple things that seem wrong or incomplete about this documentation.

1. The documentation says that class names cannot start with the string `-amp-`, but then the AMP validator doesn't actually enforce this.
2. On the other hand, the validator does enforce that class names cannot start with `i-amphtml-`, but the `i-amphtml-` prefix isn't documented anywhere (at least I didn't find any).
3. The documentation also says that you cannot reference `i-amp`, missing a `-` at the end. This is not consistent with the previous documentation which does include a `-` at the end: `i-amp-`.
4. The documentation also says that you cannot reference `i-amp-` tag names, but this is not enforced by the validator.
5. What about class names that start with `amp-` such as `amp-element` and `amp-notbuilt`? These are also used internal by the AMP runtime, but the documentation is silent about this and the validator also doesn't ban setting class names to `amp-*`. Wouldn't that cause issues with AMP trying to apply styles to classes that have `amp-` prefix?

/cc @choumx 
/cc @caroqliu

**Comment by honeybadgerdontcare** [comment]
I agree this documentation appears to be out of date.

The validator disallows class names with the following regex `(^|\W)i-amphtml-`, per [validator-main.protoascii](https://github.com/ampproject/amphtml/blob/master/validator/validator-main.protoascii#L6503). Code isn't great documentation, so hopefully we can get this updated.

@CrystalOnScript fyi

**Comment by honeybadgerdontcare** [comment]
There was a time where several variations mentioned in your post and the documentation were disallowed, but we ended up normalizing on `i-amphtml-`. Thank you for surfacing this so we can get it current.

**Comment by caroqliu** [comment]
Thanks for filing this--this issue is definitely good to reach consistency on, both in documentation and in practice. Speaking for UI `i-amphtml-` is the prefix for new features, though there are many instances from components that predate this practice. Renaming these would be a breaking change for live pages, so I wouldn't recommend it. It's also worth noting that we occasionally expose CSS classes for the explicit purpose of allowing publishers to style something, the [`autocomplete-partial`](https://amp.dev/documentation/components/amp-autocomplete/) class being one such example. 

cc @ampproject/wg-stories @ampproject/wg-ads @ampproject/wg-analytics to weigh in on their processes for exposing or restricting css classnames.

**Comment by jackbsteinberg** [comment]
Stories similarly uses the `i-amphtml-` prefix for specifically restricted CSS classes

**Comment by newmuis** [comment]
Agreed with @jackbsteinberg; just want to additionally weigh in that stories [migrated from `i-amp-` to `i-amphtml-`](https://github.com/newmuis/amphtml-story/pull/184/files) before being publicly released, so no there's no concern there.

**Comment by honeybadgerdontcare** [comment]
@dvoytenko and I worked together on unifying the internal class names to be all `i-amphtml-`, so I'm surprised to hear some are still using other prefixes. @caroqliu are you sure about this?

**Comment by caroqliu** [comment]
I should rephrase: there are instances of classes not using the `i-amphtml` pattern in our codebase without clear indication of the intention for them to be internal or exposed. 

For example the [z-index](https://github.com/ampproject/amphtml/blob/master/css/Z_INDEX.md) list includes class names not prefixed with `i-amphtml-`. [`amp-video-docking`](https://amp.dev/documentation/components/amp-video-docking/#styling) and [`amp-carousel`](https://amp.dev/documentation/components/amp-carousel/#attributes) are examples of describing that these are intended for external use, however others aren't immediately clear in terms of whether or not they are intended to be internal. i.e. a quick search reveals `.amp-date-picker-calendar-container`, '.amp-instagram-default-framing' and other 3p components.

I guess [AMP CSS classes](https://github.com/ampproject/amphtml/blob/master/spec/amp-css-classes.md) suggests that these patterns explicitly express that they are intended for publisher styling, so making the documentation on individual components using them more robust to confirm this could resolve this.

**Comment by newmuis** [comment]
> I guess AMP CSS classes suggests that these patterns explicitly express that they are intended for publisher styling, so making the documentation on individual components using them more robust to confirm this could resolve this.

Reading it this way, there are absolutely styles that we enforce in `amp-story` that apply to elements without an `i-amphtml`-prefixed classname.  AFAIK, we don't have any prefix that means "please style this," just a prefix that means "don't style this"

**Comment by choumx** [comment]
I think OP is simply asking to update documentation. Transferring to amp.dev repository which contains the [referenced page](https://github.com/ampproject/amp.dev/blob/future/pages/content/amp-dev/documentation/guides-and-tutorials/develop/style_and_layout/style_pages.md).
